engine/access/rpc/backend/transaction_subscription_metadata.go

```diff
@@ -0,0 +1,258 @@
+//
+// This metadata object is used for monitoring the transaction's progress and maintaining its state.
+//
+// Parameters:
+//   - ctx: Context to manage the operation lifecycle.
+//   - txID: The unique identifier of the transaction.
+//   - referenceBlockID: The ID of the transaction's reference block.
+//   - backendTransactions: The reference to the backend transactions module.
+//   - startHeight: The height of the block to start monitoring from.
+//   - requiredEventEncodingVersion: The required version of event encoding.
+//
+// Returns:
+//   - *transactionSubscriptionMetadata: Metadata containing the transaction's current state.
+//   - error: An error, if the metadata could not be initialized.
+func newTransactionSubscriptionMetadata(
+	ctx context.Context,
+	backendTransactions *backendTransactions,
+	txID flow.Identifier,
+	txReferenceBlockID flow.Identifier,
+	startHeight uint64,
+	eventEncodingVersion entities.EventEncodingVersion,
+) (*transactionSubscriptionMetadata, error) {
+	txMetadata := &transactionSubscriptionMetadata{
+		backendTransactions:  backendTransactions,
+		TransactionResult:    &access.TransactionResult{TransactionID: txID},
+		txReferenceBlockID:   txReferenceBlockID,
+		eventEncodingVersion: eventEncodingVersion,
+	}
+
+	if err := txMetadata.initTransactionReferenceBlockID(); err != nil {
+		return nil, err
+	}
+
+	if err := txMetadata.initBlockInfoFromBlocksRange(startHeight); err != nil {
+		return nil, err
+	}
+
+	if err := txMetadata.initTransactionResult(ctx); err != nil {
+		return nil, err
+	}
+
+	return txMetadata, nil
+}
+
+// initTransactionReferenceBlockID initializes the reference block ID for the transaction.
+// If the reference block ID is not set, it fetches the transaction details and updates the reference block ID.
+//
+// Returns:
+//   - error: An error, if the reference block ID cannot be determined.
+func (tm *transactionSubscriptionMetadata) initTransactionReferenceBlockID() error {
+	// Get referenceBlockID if it is not set
+	if tm.txReferenceBlockID == flow.ZeroID {
+		tx, err := tm.backendTransactions.transactions.ByID(tm.TransactionID)
+		if err != nil {
+			return err
+		}
+		tm.txReferenceBlockID = tx.ReferenceBlockID
+	}
+
+	return nil
+}
+
+// initBlockInfoFromBlocksRange initializes block information for the transaction by searching through a range of blocks.
+//
+// This function iterates over blocks from the transaction's reference block height to a specified end height,
+// searching for the block containing the transaction.
+//
+// Parameters:
+//   - end: The ending block height to stop the search.
+//
+// Returns:
+//   - error: An error, if the block information cannot be initialized.
+func (tm *transactionSubscriptionMetadata) initBlockInfoFromBlocksRange(end uint64) error {
+	refBlock, err := tm.backendTransactions.blocks.ByID(tm.txReferenceBlockID)
+	if err != nil {
+		return err
+	}
+
+	for height := refBlock.Header.Height; height <= end; height++ {
```

Comment:
I don't think we should need to iterate over all of the blocks. if this tx wasn't just sent, we can use similar logic as the polling endpoint to see if we have the data. Then, use the current sealed block as the starting height.

https://github.com/onflow/flow-go/blob/c396112c7c06ab8da2353381c04e2cc3c5177ffc/engine/access/rpc/backend/backend_transactions.go#L630-L642